jerry-core/parser/js/js-parser-module.c

```diff
@@ -0,0 +1,735 @@
+#include "js-parser-internal.h"
+#include "jerryscript-port.h"
+#include "jcontext.h"
+
+#include "ecma-exceptions.h"
+#include "ecma-function-object.h"
+#include "ecma-helpers.h"
+#include "ecma-literal-storage.h"
+#include "ecma-lex-env.h"
+
+#ifndef CONFIG_DISABLE_ES2015_MODULE_SYSTEM
+static bool
+parser_module_check_for_duplicates_in_node (parser_module_node_t *module_node_p,
+                                            const uint8_t *import_name_p,
+                                            prop_length_t import_name_length)
+{
+  parser_module_names_t *import_names_p = module_node_p->module_names;
+
+  while (import_names_p != NULL)
+  {
+    uint8_t *current_p = import_names_p->import_name_p;
+    prop_length_t current_length = import_names_p->import_name_length;
+
+    if (current_p != NULL && current_length == import_name_length)
+    {
+      if (memcmp (current_p, import_name_p, current_length) == 0)
+      {
+        return true;
+      }
+    }
+
+    import_names_p = import_names_p->next;
+  }
+  return false;
+} /* parser_module_check_for_duplicates_in_node */
+
+static bool
+parser_module_check_for_duplicates (parser_context_t *context_p,
+                                    parser_module_node_t *module_node_p,
+                                    lexer_literal_t *import_name_p)
+{
+  if (import_name_p == NULL)
+  {
+    return false;
+  }
+
+  bool hasDuplicate = parser_module_check_for_duplicates_in_node (module_node_p,
+                                                                  import_name_p->u.char_p,
+                                                                  import_name_p->prop.length);
+
+  if (!hasDuplicate)
+  {
+    parser_module_node_t *node_p = context_p->module_context_p->imports;
+    while (node_p != NULL && !hasDuplicate)
+    {
+      hasDuplicate = parser_module_check_for_duplicates_in_node (node_p,
+                                                                 import_name_p->u.char_p,
+                                                                 import_name_p->prop.length);
+      node_p = node_p->next;
+    }
+  }
+
+  return hasDuplicate;
+} /* parser_module_check_for_duplicates */
+
+static void
+parser_module_delete_saved_names (parser_module_names_t *module_names_p, uint8_t name_count)
+{
+  parser_module_names_t *current_p = module_names_p;
+  parser_module_names_t *next_p;
+
+  for (int i = 0; i < name_count; i++)
+  {
+    next_p = current_p->next;
+    parser_free (current_p->import_name_p, current_p->import_name_length * sizeof (uint8_t));
+    parser_free (current_p->local_name_p, current_p->local_name_length * sizeof (uint8_t));
+    parser_free (current_p, sizeof (parser_module_names_t));
+    current_p = next_p;
+  }
+} /* parser_module_delete_saved_names */
+
+static void
+parser_module_partial_cleanup_on_error (parser_module_node_t *module_node_p)
+{
+  if (module_node_p != NULL)
+  {
+    parser_module_delete_saved_names (module_node_p->module_names,
```

Comment:
We could simply pass the `module_node_p` directly and accessing the `names`, `module_request_count` can be done in that method.